integrations/serverpod_cloud_storage_gcp/lib/src/google_cloud_storage.dart

```diff
@@ -0,0 +1,289 @@
+import 'dart:collection';
+import 'dart:convert';
+import 'dart:io';
+import 'dart:typed_data';
+
+import 'package:crypto/crypto.dart';
+import 'package:hex/hex.dart';
+import 'package:gcloud/storage.dart';
+import 'package:googleapis_auth/auth_io.dart' as auth;
+// ignore: implementation_imports
+import 'package:googleapis_auth/src/crypto/pem.dart';
+// ignore: implementation_imports
+import 'package:googleapis_auth/src/crypto/rsa_sign.dart';
+import 'package:intl/intl.dart';
+import 'package:serverpod/serverpod.dart';
+
+/// Concrete implementation of Google Cloud Storage, using native GCP APIs,
+/// for use with Serverpod.
+class GoogleCloudStorage extends CloudStorage {
+  final String bucket;
+  final bool public;
+
+  late final String publicHost;
+  late final Storage storage;
+
+  /// The path to service account credentials with storage admin permissions
+  final String serviceAccountPath;
+  late final Map<String, dynamic> _serviceAccount;
+
+  /// Creates a new [GoogleCloudStorage] instance.
+  /// Uses service account credentials to authenticate to Cloud Storage.
+  ///
+  /// Usage:
+  ///   final publicStorage = await GoogleCloudStorage.init(
+  ///     serverpod: pod,
+  ///     storageId: 'public',
+  ///     public: true,
+  ///     bucket: myBucket,
+  ///     serviceAccountPath: '/config/google_account_credentials.json',
+  ///   );
+  ///
+  ///   pod.addCloudStorage(publicStorage);
+  static Future<GoogleCloudStorage> init({
+    required Serverpod serverpod,
+    required String storageId,
+    required bool public,
+    required String bucket,
+    required String serviceAccountPath,
```

Comment:
Suggestion: 

This comment comes with a few implications. By passing in a path to the service account we are forcing people to use a service account stored on disc to authenticate with google. But there are otherways people may want to auth such as via the metaserver in a hosted environment. Google also has a method for default authentication, e.g. `clientViaApplicationDefaultCredentials` that will use an environment variable with the path to the service account, or the metaserver if running in cloud engine. 

So it would be nice to use this method instead of passing in the path to the service account.

On top of this, it would be great to take an optional http client. If set it would override the call to `clientViaApplicationDefaultCredentials`. Then the user can use any available Google auth client or even a mock client for testing.

There is one implication of this though, and that is that we are currently signing the URL in process because we have direct access to the private key. This is not possible if we have authed via the meta server. 

Luckily Google has an API to sign the URL though, but it requires the service account email. We could in other words take the service account email as input instead of the service account path and call this endpoint.

Like so:

```dart
var _iamCredentials = IAMCredentialsApi(authedClient);

final response = await _iamCredentials.projects.serviceAccounts.signBlob(
  SignBlobRequest(payload: base64.encode(stringToSignList)),
  'projects/-/serviceAccounts/$_serviceAccountEmail',
);
```

Where the the class `IamCredentials` comes from `import 'package:googleapis/iamcredentials/v1.dart';`

With the above change we can remove all service account management inside the code.

Reply:
I think using the API to get a signature makes sense, and removes the burden of maintaining the signature logic. I'll update it